### ISSUE TITLE
[css-grid-3] FPWD for Masonry Layout

### ISSUE THREAD

**Issue by jensimmons** [w3c/csswg-drafts#8195]
There is a draft for Masonry layout at https://drafts.csswg.org/css-grid-3/

We would like to see this go to First Public Working Draft.

**Comment by bfgeek** [comment]
We quickly reviewed the current state of the specification. There are a few concerns we have with the general direction (to be clear - not the use-cases, and desire for the feature in general - just the current direction of the specification).

1. Issues with intrinsic sizing of masonry grids. https://github.com/w3c/csswg-drafts/issues/8206

Grid works by placing everything in the grid ahead of time, then sizing the rows/columns to fit the items. Masonry fundamentally doesn't work this way as you need to size the rows/columns ahead of time - then place items within those rows/columns.

As a result the way the current specification re-uses the grid sizing logic leads to poor results when intrinsically sizing tracks, and if the grid is intrinsically-sized itself (e.g. if its within a grid/flex/table, etc).

Here is a simple example demonstrating one aspect of this issue within Firefox Nightly. https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11094

How to solve this particular issue is a relatively fundamental issue for this specification. There are many different potential avenues for solving this. From masonry examples we've seen in particular there is a desire to have repeaters which are intrinsically sized (which isn't currently possible with regular grid repeaters). We suspect that re-using the grid algorithm as described isn't the best approach here.

As-is its unclear if attempting to re-use the grid placement & sizing logic results in the best feature for web developers.

2. Issues with align-tracks/justify-tracks. https://github.com/w3c/csswg-drafts/issues/8207 and https://github.com/w3c/csswg-drafts/issues/8208

As-is this sub-feature has a few issues. The primary concern is with accessibility. This feature allows authors to create order with unclear visual ordering (e.g. even with the new `visual-order` feature we'd still have issues). From examples of masonry layouts we've seen - there doesn't seem to be much desire for aligning indivudal tracks, rather the primary desire is switching the direction for the "start" mansonry edge which the current specification doesn't address.

There is a similar accessibility concern with `masonry-auto-flow: next`.

Additionally this sub-feature doesn't have any logic for dealing with items spanning tracks (in almost every non-start case leading to undesirable overlapping content in our testing within Firefox Nightly).

We suspect there is an alternate solution space here which doesn't involve aligning individual tracks. E.g. masonry.js has flags to flip the "origin" https://masonry.desandro.com/options.html#origintop There is likely a algorithm that could be used for stretching all the tracks, in such a way that it keeps the visual order consistent.

We'd like to see these class of issues thought about before moving to FPWD.

Ian

**Comment by jensimmons** [comment]
>From examples of masonry layouts I've seen - there doesn't seem to be much desire for aligning individual tracks, rather the primary desire is switching the direction for the "start" masonry edge which the current specification doesn't address.

>I suspect there is an alternate solution space here which doesn't involve aligning individual tracks. 

I agree. At first glance, I'm not sure what the compelling use case is for aligning individual tracks. It seems far more common designers/developers would want to change the alignment of all the tracks from start to end, etc. Not each separately.

**Comment by henke37** [comment]
I think that it would be a good addition to have a non-normative section that compares the feature against the dense mode of [grid-auto-flow](https://drafts.csswg.org/css-grid-2/#propdef-grid-auto-flow).

**Comment by svgeesus** [comment]
It seems there is some support for the concerns raised by @bfgeek and thus, unless the editors now think otherwise, I feel that calling for FPWD is premature.

@tabatkins @fantasai @MatsPalmgren @jensimmons @stwrt 

(I'm normally all for early-enough FPWD, but in this specific case where the overall direction is in question it seems premature).

I suggest removing the Agenda+ and, once there is rough consensus on the overall direction, re-applying it.

Linking other relevant masonry threads to this one would also probably help, including (especially) once they are closed ie there is agreement.

**Comment by chrishtr** [comment]
Removing Agenda+ since we can't discuss this issue until there is more discussion on #9041. We should discuss that first.

**Comment by fantasai** [comment]
@tabatkins and I worked through their [Alternate Masonry Draft](https://tabatkins.github.io/specs/css-masonry/#repeat) (#9041) as well as the Needs Edits issues against masonry and merged everything into the current [CSS Grid Layout Level 3 Editor's Draft](https://drafts.csswg.org/css-grid-3/#alignment).

The changes include:
* Outlining both syntax proposals in parallel, while factoring out everything that remains common. Sections that are different are called out with colored bars on the right edge.
* Sketching out the [optimized version](https://drafts.csswg.org/css-grid-3/#track-sizing-performance) of the track sizing algorithm, including [special handling of submasonry items](https://drafts.csswg.org/css-grid-3/#track-sizing-subgrid).
* Incorporating the `masonry-slack` proposal to [specify tie thresholds](https://drafts.csswg.org/css-grid-3/#masonry-slack) as resolved in #9328.
* Giving `auto-fit` [rules for collapsing “empty” tracks](https://drafts.csswg.org/css-grid-3/#repeat-auto-fit) per #10274.
* Adding `row-reverse`, `column-reverse`, and `wrap-reverse` keywords to `grid-auto-flow` per #3622 and applying them to masonry placement in the grid-integrated model. (The grid-independent model uses `masonry-flow`.)
* Introducing an `auto-areas` repeat keyword that generates a size listing to match the number of tracks created by template areas, for both grid and masonry containers. (This was a new feature in Tab's draft.)

We'd like to request FPWD of this combined draft to encourage review and discussion by the broader community.

**Comment by fantasai** [comment]
We also added a new [accessibility considerations section](https://drafts.csswg.org/css-grid-3/#order-accessibility).
